### ISSUE TITLE
Shader execution support

### ISSUE THREAD

**Issue by DemiMarie** [wgslfuzz/darthshader#3]
It is possible to not only compile shaders, but also to actually execute the compiled code (which, after all, is the whole point!).  It would be interesting to find bugs that do not result in shader compiler crashes, but which instead result in bad code being generated.  This bad code might itself perform insecure operations, such as accessing out of bounds data.

**Comment by wgslfuzz** [comment]
Out-of-bound access on the GPU, I presume? Detecting those would be interesting, as this could expose not only bugs due to bad code generation in the shader compiler but also subsequent stages. Actually executing the generated code would require plenty of work on the harness-side + a proper fault oracle. Since we don't know the expected output of a fuzzer-generated shader we might need to resort to differential testing (e.g., initialize the GPU memory with patternX, run shader + extract outputs, fill GPU memory with patternY, run shader, compare the outputs with the outputs for the previous execution).

**Comment by DemiMarie** [comment]
What about differential testing between different GPUs?

**Comment by wgslfuzz** [comment]
Differential testing, either between different GPUs or maybe a single GPU and SwiftShader could result in some differential behavior. This begs the question: what's the root cause and are there security implications. One scenario could be leaking GPU memory => differential behavior with security impact.
One problem come to mind with this differential testing approach:  shaders might contain undefined behavior (as in, floating point madness). This will result in differentials not actually being meaningful. Haven't thought about this, maybe having 3 implementations could help? If all 3 differ => floating-point madness. If two agree and a third one diverges => investigate.

**Comment by DemiMarie** [comment]
Also, if none of the shaders use floating point then any discrepency should be investigated.

**Comment by wgslfuzz** [comment]
Indeed, this is a good point. The fuzzer could be adapted to avoid floating-points altogether and only compute integer results.

**Comment by DemiMarie** [comment]
I think floating point inputs should be considered, as they will trigger code paths (such as optimizations) that otherwise would not be triggered.
